File: ios/Utils/ErrorMapper.m
//
//  ErrorMapper.m
//  HuaweiObs
//
//  Created by react-native-huawei-obs
//

#import "ErrorMapper.h"
#import <OBS/OBS.h>

@implementation ErrorMapper

+ (NSDictionary *)mapError:(NSError *)error {
    // OBS SDK 服务端错误 (com.obs.serverError)
    if ([error.domain isEqualToString:@"com.obs.serverError"]) {
        return [self mapOBSServerError:error];
    }
    
    // OBS SDK 客户端错误 (com.obs.services.error)
    if ([error.domain isEqualToString:@"com.obs.services.error"]) {
        return [self mapOBSError:error];
    }
    
    // URL 错误
    if ([error.domain isEqualToString:NSURLErrorDomain]) {
        return [self mapURLError:error];
    }
    
    // 文件错误
    if ([error.domain isEqualToString:NSCocoaErrorDomain]) {
        return [self mapFileError:error];
    }
    
    // 自定义错误
    NSString *code = error.userInfo[@"code"];
    if (code) {
        return @{
            @"code": code,
            @"message": error.localizedDescription,
            @"isRetryable": @([self isRetryableErrorCode:code])
        };
    }
    
    // 未知错误
    return @{
        @"code": @"E_UNKNOWN",
        @"message": error.localizedDescription ?: @"Unknown error",
        @"isRetryable": @NO
    };
}

+ (NSDictionary *)mapOBSServerError:(NSError *)error {
    // com.obs.serverError 格式：
    // userInfo 包含 statusCode, requestID, responseXMLBodyDict (含 Code, Message, RequestId)
    NSDictionary *xmlBody = error.userInfo[@"responseXMLBodyDict"];
    NSString *errorCode = xmlBody[@"Code"] ?: @"";
    NSNumber *statusCode = error.userInfo[@"statusCode"] ?: @0;
    NSString *requestId = error.userInfo[@"xRequestID"] ?: xmlBody[@"RequestId"] ?: @"";
    NSString *serverMessage = xmlBody[@"Message"] ?: error.localizedDescription ?: @"OBS server error";
    
    NSString *code;
    BOOL isRetryable = NO;
    
    // Map OBS error codes
    if ([errorCode isEqualToString:@"InvalidAccessKeyId"]) {
        code = @"E_AUTH_INVALID_CREDENTIAL";
    } else if ([errorCode isEqualToString:@"SignatureDoesNotMatch"]) {
        code = @"E_AUTH_SIGNATURE_MISMATCH";
    } else if ([errorCode isEqualToString:@"RequestTimeTooSkewed"]) {
        code = @"E_AUTH_TIME_SKEWED";
    } else if ([errorCode isEqualToString:@"AccessDenied"]) {
        code = @"E_AUTH_ACCESS_DENIED";
    } else if ([errorCode isEqualToString:@"ExpiredToken"]) {
        code = @"E_AUTH_EXPIRED";
    } else if ([errorCode isEqualToString:@"NoSuchBucket"]) {
        code = @"E_BUCKET_NOT_FOUND";
    } else if ([errorCode isEqualToString:@"NoSuchKey"]) {
        code = @"E_FILE_NOT_FOUND";
    } else if ([errorCode isEqualToString:@"TooManyRequests"]) {
        code = @"E_CONCURRENCY_LIMIT_EXCEEDED";
        isRetryable = YES;
    } else if ([errorCode isEqualToString:@"RequestTimeout"]) {
        code = @"E_NETWORK_TIMEOUT";
        isRetryable = YES;
    } else if ([errorCode isEqualToString:@"InternalError"] ||
               [errorCode isEqualToString:@"ServiceUnavailable"]) {
        code = @"E_HTTP_5XX";
        isRetryable = YES;
    } else {
        code = [self mapHTTPStatus:[statusCode intValue]];
        isRetryable = [self isRetryableHTTPStatus:[statusCode intValue]];
    }
    
    return @{
        @"code": code,
        @"message": [NSString stringWithFormat:@"[%@] %@", errorCode, serverMessage],
        @"statusCode": statusCode,
        @"requestId": requestId,
        @"isRetryable": @(isRetryable)
    };
}

+ (NSDictionary *)mapOBSError:(NSError *)error {
    NSString *errorCode = error.userInfo[@"ErrorCode"] ?: @"";
    NSNumber *statusCode = error.userInfo[@"StatusCode"] ?: @0;
    NSString *requestId = error.userInfo[@"RequestId"] ?: @"";
    NSString *hostId = error.userInfo[@"HostId"] ?: @"";
    
    NSString *code;
    BOOL isRetryable = NO;
    
    // Map OBS error codes
    if ([errorCode isEqualToString:@"InvalidAccessKeyId"]) {
        code = @"E_AUTH_INVALID_CREDENTIAL";
    } else if ([errorCode isEqualToString:@"SignatureDoesNotMatch"]) {
        code = @"E_AUTH_SIGNATURE_MISMATCH";
    } else if ([errorCode isEqualToString:@"RequestTimeTooSkewed"]) {
        code = @"E_AUTH_TIME_SKEWED";
    } else if ([errorCode isEqualToString:@"AccessDenied"]) {
        code = @"E_AUTH_ACCESS_DENIED";
    } else if ([errorCode isEqualToString:@"ExpiredToken"]) {
        code = @"E_AUTH_EXPIRED";
    } else if ([errorCode isEqualToString:@"NoSuchBucket"]) {
        code = @"E_BUCKET_NOT_FOUND";
    } else if ([errorCode isEqualToString:@"NoSuchKey"]) {
        code = @"E_FILE_NOT_FOUND";
    } else if ([errorCode isEqualToString:@"TooManyRequests"]) {
        code = @"E_CONCURRENCY_LIMIT_EXCEEDED";
        isRetryable = YES;
    } else if ([errorCode isEqualToString:@"RequestTimeout"]) {
        code = @"E_NETWORK_TIMEOUT";
        isRetryable = YES;
    } else if ([errorCode isEqualToString:@"InternalError"] || 
               [errorCode isEqualToString:@"ServiceUnavailable"]) {
        code = @"E_HTTP_5XX";
        isRetryable = YES;
    } else {
        // Map by HTTP status code
        code = [self mapHTTPStatus:[statusCode intValue]];
        isRetryable = [self isRetryableHTTPStatus:[statusCode intValue]];
    }
    
    return @{
        @"code": code,
        @"message": error.localizedDescription ?: @"OBS error",
        @"statusCode": statusCode,
        @"requestId": requestId,
        @"hostId": hostId,
        @"isRetryable": @(isRetryable)
    };
}

+ (NSDictionary *)mapURLError:(NSError *)error {
    NSString *code;
    BOOL isRetryable = NO;
    
    switch (error.code) {
        case NSURLErrorTimedOut:
            code = @"E_NETWORK_TIMEOUT";
            isRetryable = YES;
            break;
        case NSURLErrorNotConnectedToInternet:
        case NSURLErrorNetworkConnectionLost:
            code = @"E_NETWORK_CONNECTION_LOST";
            isRetryable = YES;
            break;
        case NSURLErrorDNSLookupFailed:
            code = @"E_NETWORK_DNS_FAILED";
            isRetryable = YES;
            break;
        case NSURLErrorCannotFindHost:
            code = @"E_NETWORK_HOST_UNREACHABLE";
            isRetryable = YES;
            break;
        case NSURLErrorCancelled:
            code = @"E_TASK_CANCELLED";
            break;
        default:
            code = @"E_NETWORK_REQUEST_FAILED";
            isRetryable = YES;
            break;
    }
    
    return @{
        @"code": code,
        @"message": error.localizedDescription ?: @"Network error",
        @"isRetryable": @(isRetryable)
    };
}

+ (NSDictionary *)mapFileError:(NSError *)error {
    NSString *code;
    
    switch (error.code) {
        case NSFileNoSuchFileError:
            code = @"E_FILE_NOT_FOUND";
            break;
        case NSFileReadNoPermissionError:
            code = @"E_FILE_NOT_READABLE";
            break;
        case NSFileWriteNoPermissionError:
            code = @"E_FILE_NOT_WRITABLE";
            break;
        case NSFileWriteOutOfSpaceError:
            code = @"E_FILE_NO_SPACE";
            break;
        default:
            code = @"E_FILE_IO_ERROR";
            break;
    }
    
    return @{
        @"code": code,
        @"message": error.localizedDescription ?: @"File error",
        @"isRetryable": @NO
    };
}

+ (NSString *)mapHTTPStatus:(NSInteger)statusCode {
    switch (statusCode) {
        case 400:
            return @"E_HTTP_BAD_REQUEST";
        case 401:
            return @"E_AUTH_INVALID_CREDENTIAL";
        case 403:
            return @"E_AUTH_ACCESS_DENIED";
        case 404:
            return @"E_FILE_NOT_FOUND";
        case 409:
            return @"E_HTTP_CONFLICT";
        case 429:
            return @"E_CONCURRENCY_LIMIT_EXCEEDED";
        case 500 ... 599:
            return @"E_HTTP_5XX";
        default:
            return @"E_HTTP_UNKNOWN";
    }
}

+ (BOOL)isRetryableHTTPStatus:(NSInteger)statusCode {
    return statusCode >= 500 || statusCode == 429;
}

+ (BOOL)isRetryableErrorCode:(NSString *)code {
    NSSet *retryableCodes = [NSSet setWithArray:@[
        @"E_NETWORK_TIMEOUT",
        @"E_NETWORK_CONNECTION_LOST",
        @"E_NETWORK_REQUEST_FAILED",
        @"E_HTTP_5XX",
        @"E_CONCURRENCY_LIMIT_EXCEEDED",
        @"E_UPLOAD_PART_FAILED"
    ]];
    return [retryableCodes containsObject:code];
}

@end

File: android/src/main/java/com/huaweiobs/utils/ErrorMapper.java
package com.huaweiobs.utils;

import com.obs.services.exception.ObsException;

import java.io.IOException;
import java.net.SocketTimeoutException;
import java.net.UnknownHostException;
import java.util.Arrays;
import java.util.List;

/**
 * 错误映射器
 * 将 OBS SDK 异常映射为统一的错误码
 */
public class ErrorMapper {

    /**
     * 映射异常到 OBS 错误
     */
    public static OBSException mapException(Throwable e) {
        if (e instanceof OBSException) {
            return (OBSException) e;
        } else if (e instanceof ObsException) {
            return mapObsException((ObsException) e);
        } else if (e instanceof SocketTimeoutException) {
            OBSException exception = new OBSException(
                "E_NETWORK_TIMEOUT",
                "Network timeout: " + e.getMessage(),
                e
            );
            exception.setRetryable(true);
            return exception;
        } else if (e instanceof UnknownHostException) {
            OBSException exception = new OBSException(
                "E_NETWORK_UNAVAILABLE",
                "Network unavailable: " + e.getMessage(),
                e
            );
            exception.setRetryable(true);
            return exception;
        } else if (e instanceof IOException) {
            OBSException exception = new OBSException(
                "E_NETWORK_ERROR",
                "Network error: " + e.getMessage(),
                e
            );
            exception.setRetryable(true);
            return exception;
        } else if (e instanceof IllegalArgumentException) {
            return new OBSException(
                "E_INVALID_ARGUMENT",
                e.getMessage() != null ? e.getMessage() : "Invalid argument",
                e
            );
        } else {
            return new OBSException(
                "E_UNKNOWN",
                e.getMessage() != null ? e.getMessage() : "Unknown error",
                e
            );
        }
    }

    /**
     * 映射 OBS 官方 SDK 异常
     */
    private static OBSException mapObsException(ObsException e) {
        String errorCode = e.getErrorCode();
        int responseCode = e.getResponseCode();
        String message = e.getErrorMessage() != null ? e.getErrorMessage() : 
                        (e.getMessage() != null ? e.getMessage() : "Unknown OBS error");

        String sdkCode;

        // 认证错误
        if ("InvalidAccessKeyId".equals(errorCode) || "InvalidSecretAccessKey".equals(errorCode)) {
            sdkCode = "E_AUTH_INVALID_CREDENTIAL";
        } else if ("AccessDenied".equals(errorCode)) {
            sdkCode = "E_AUTH_ACCESS_DENIED";
        } else if ("SignatureDoesNotMatch".equals(errorCode)) {
            sdkCode = "E_AUTH_SIGNATURE_MISMATCH";
        } else if ("TokenExpired".equals(errorCode) || "RequestTimeTooSkewed".equals(errorCode)
                || "SecurityTokenExpired".equals(errorCode)) {
            sdkCode = "E_AUTH_EXPIRED";
        } else if ("InvalidToken".equals(errorCode) || "AuthFailure".equals(errorCode)
                || "SecurityTokenMalformed".equals(errorCode) || "InvalidStsAccessKeyId".equals(errorCode)) {
            sdkCode = "E_AUTH_INVALID_CREDENTIAL";
        }
        // 资源不存在
        else if ("NoSuchBucket".equals(errorCode)) {
            sdkCode = "E_BUCKET_NOT_FOUND";
        } else if ("NoSuchKey".equals(errorCode)) {
            sdkCode = "E_FILE_NOT_FOUND";
        } else if ("NoSuchUpload".equals(errorCode)) {
            sdkCode = "E_UPLOAD_NOT_FOUND";
        }
        // 网络相关
        else if ("RequestTimeout".equals(errorCode)) {
            sdkCode = "E_NETWORK_TIMEOUT";
        } else if ("ServiceUnavailable".equals(errorCode)) {
            sdkCode = "E_HTTP_5XX";
        } else if ("InternalError".equals(errorCode)) {
            sdkCode = "E_HTTP_5XX";
        }
        // 根据 HTTP 状态码映射
        else if (responseCode >= 500) {
            sdkCode = "E_HTTP_5XX";
        } else if (responseCode == 429) {
            sdkCode = "E_CONCURRENCY_LIMIT_EXCEEDED";
        } else if (responseCode == 408) {
            sdkCode = "E_NETWORK_TIMEOUT";
        } else if (responseCode == 404) {
            sdkCode = "E_FILE_NOT_FOUND";
        } else if (responseCode == 403) {
            sdkCode = "E_AUTH_ACCESS_DENIED";
        } else if (responseCode == 401) {
            sdkCode = "E_AUTH_INVALID_CREDENTIAL";
        } else if (responseCode >= 400) {
            sdkCode = "E_HTTP_4XX";
        } else {
            sdkCode = "E_UNKNOWN";
        }

        // 在消息中加入 OBS 错误码前缀，方便调试
        String fullMessage = errorCode != null && !errorCode.isEmpty()
            ? "[" + errorCode + "] " + message
            : message;

        OBSException exception = new OBSException(sdkCode, fullMessage, e);
        exception.setStatusCode(responseCode);
        exception.setRequestId(e.getErrorRequestId());
        exception.setHostId(e.getErrorHostId());
        exception.setRetryable(isRetryableErrorCode(sdkCode));
        return exception;
    }

    /**
     * 判断错误是否可重试
     */
    private static boolean isRetryableErrorCode(String code) {
        List<String> retryableCodes = Arrays.asList(
            "E_NETWORK_TIMEOUT",
            "E_NETWORK_ERROR",
            "E_NETWORK_UNAVAILABLE",
            "E_HTTP_5XX",
            "E_UPLOAD_PART_FAILED"
        );
        return retryableCodes.contains(code);
    }
}
